rz-svd/rz-svd.c

```diff
@@ -0,0 +1,309 @@
+	return *ta->ptr == 0 || ta->ptr >= ta->end;
+}
+
+static ta_iter *ta_iter_next(ta_iter *ta) {
+	yxml_ret_t r = YXML_OK;
+	yxml_t ta_x;
+
+	int level;
+	char *cur, *tmp;
+	char value[VALUESIZE];
+	const char *ta_start;
+
+	cur = value;
+	value[0] = 0;
+
+	if (!ta->baseaddress[0]) {
+		while (!ta_iter_done(ta) &&
+			(yxml_parse(&ta->x, *ta->ptr) != YXML_ELEMSTART ||
+				strcasecmp(ta->x.elem, "peripherals"))) {
+			ta->ptr++;
+		}
+		if (ta_iter_done(ta))
+			return NULL;
+		ta_start = ta->ptr;
+		ta_x = ta->x;
+
+		level = 0;
+		while (!ta_iter_done(ta) && !ta->baseaddress[0]) {
+			switch ((r = yxml_parse(&ta->x, *ta->ptr))) {
+			case YXML_ELEMSTART:
+				level += 1;
+				if (level == 2 && strcasecmp(ta->x.elem, "baseAddress") == 0) {
+					cur = value;
+					*cur = 0;
+				}
+				break;
+
+			case YXML_ELEMEND:
+				level -= 1;
+				if (level < 0) {
+					return ta_iter_next(ta);
+				} else if (level == 2 && cur) {
+					strcpytrunc(ta->baseaddress, value, sizeof(ta->baseaddress));
+					cur = NULL;
+
+					// go back to the beginning of peripherals?
+					ta->ptr = ta_start;
+					ta->x = ta_x;
+				}
+				break;
+
+			case YXML_CONTENT:
+				if (!cur || level != 2) {
+					break;
+				}
+				tmp = ta->x.data;
+				while (*tmp && cur < value + sizeof(value)) {
+					*cur++ = *tmp++;
+				}
+				if (cur >= value + sizeof(value)) {
+					cur = NULL;
+				} else {
+					*cur = 0;
+				}
+				break;
+			default:
+				break;
+			}
+			ta->ptr++;
+		}
+		if (ta_iter_done(ta))
+			return NULL;
+	}
+	assert(ta->baseaddress[0]);
+
+	// find <register>
+	level = 0;
+	while (!ta_iter_done(ta)) {
+		r = yxml_parse(&ta->x, *ta->ptr);
+
+		if (r == YXML_ELEMSTART) {
+			level += 1;
+			if (level == 3 &&
+				strcasecmp(ta->x.elem, "register") == 0) {
+				break;
+			}
+
+		} else if (r == YXML_ELEMEND) {
+			level -= 1;
+			if (level < 1) {
+				ta->baseaddress[0] = 0;
+				return ta_iter_next(ta);
+			}
+		}
+		ta->ptr++;
+	}
+	if (ta_iter_done(ta)) {
+		return NULL;
+	}
+
+	assert(r == YXML_ELEMSTART);
+
+	cur = NULL;
+	value[0] = 0;
+	elem_type = -1;
+
+	*ta->bitoffset = *ta->regname = *ta->bitwidth = 0;
+
+	for (;;) {
+		switch (r) {
+		case YXML_ELEMSTART:
+			level += 1;
+			if (level != 5)
+				break;
+			if (strcasecmp(ta->x.elem, "displayName") == 0) {
+				elem_type = REGNAME;
+			} else if (strcasecmp(ta->x.elem, "description") == 0) {
+				elem_type = DESCRIPTION;
+			} else if (strcasecmp(ta->x.elem, "addressOffset") == 0) {
+				elem_type = BITOFFSET;
+			} else if (strcasecmp(ta->x.elem, "size") == 0) {
+				elem_type = BITWIDTH;
+			} else {
+				break;
+			}
+			cur = value;
+			*cur = 0;
+			break;
+
+		case YXML_ELEMEND:
+			level -= 1;
+			if (level < 2) {
+				ta->baseaddress[0] = 0;
+				return ta_iter_next(ta);
+			} else if (level != 4 || !cur) {
+				break;
+			}
+			cur = NULL;
+			switch (elem_type) {
+			case BITOFFSET:
+				strcpytrunc(ta->bitoffset, value, sizeof(ta->bitoffset));
```

Comment:
I think this function is not available everywhere. Please substitute with functions from `rz_util`.

Reply:
It's defined here in the same file, actually: https://github.com/rizinorg/rizin-extras/pull/4/files#diff-784b07645712e153267bd55ec4e84f60a5038d9305822b974f4040b52060722fR35-R41

I'll see if there's anything in `rz_util`, as well.

Reply:
I think `rz_str_ncpy()` does that.

Reply:
By the way, the `rz_str_ncpy()` implementation looks inefficient, might make sense to make a pull request to use `memcpy` in it.

Reply:
Yup! Done.